packages/remix-dev/cli/convert-to-javascript.ts

```diff
@@ -0,0 +1,91 @@
+import path from "path";
+import fse from "fs-extra";
+import glob from "fast-glob";
+import * as babel from "@babel/core";
+// @ts-expect-error these modules dont have types
+import babelPluginSyntaxJSX from "@babel/plugin-syntax-jsx";
+// @ts-expect-error these modules dont have types
+import babelPresetTypeScript from "@babel/preset-typescript";
+import prettier from "prettier";
+
+function convertToJavaScript(
+  filename: string,
+  source: string,
+  projectDir: string
+): string {
+  let result = babel.transformSync(source, {
+    filename,
+    presets: [[babelPresetTypeScript, { jsx: "preserve" }]],
+    plugins: [babelPluginSyntaxJSX],
+    compact: false,
+    retainLines: true,
+    cwd: projectDir,
+  });
+
+  if (!result || !result.code) {
+    throw new Error("Could not parse typescript");
+  }
+
+  /*
+    Babel's `compact` and `retainLines` options are both bad at formatting code.
+    Use Prettier for nicer formatting.
+  */
+  return prettier.format(result.code, { parser: "babel" });
+}
+
+export async function convertTemplateToJavaScript(projectDir: string) {
+  // 1. Convert all .ts files in the template to .js
+  let entries = glob.sync("**/*.+(ts|tsx)", {
+    cwd: projectDir,
+    absolute: true,
+    ignore: ["**/node_modules/**"],
+  });
+  for (let entry of entries) {
+    if (entry.endsWith(".d.ts")) {
+      fse.removeSync(entry);
+      continue;
+    }
+
+    let contents = fse.readFileSync(entry, "utf8");
+    let filename = path.basename(entry);
+    let javascript = convertToJavaScript(filename, contents, projectDir);
+
+    fse.writeFileSync(entry, javascript, "utf8");
+    if (entry.endsWith(".tsx")) {
+      fse.renameSync(entry, entry.replace(/\.tsx?$/, ".jsx"));
+    } else {
+      fse.renameSync(entry, entry.replace(/\.ts?$/, ".js"));
+    }
+  }
+
+  // 2. Rename the tsconfig.json to jsconfig.json
+  if (fse.existsSync(path.join(projectDir, "tsconfig.json"))) {
+    fse.renameSync(
+      path.join(projectDir, "tsconfig.json"),
+      path.join(projectDir, "jsconfig.json")
+    );
+  }
+
+  // 3. Remove @types/* and typescript from package.json
+  let packageJson = path.join(projectDir, "package.json");
+  if (!fse.existsSync(packageJson)) {
+    throw new Error("Could not find package.json");
+  }
+  let pkg = JSON.parse(fse.readFileSync(packageJson, "utf8"));
+  let devDeps = pkg.devDependencies || {};
+  let newPackageJson = {
+    ...pkg,
+    devDependencies: Object.fromEntries(
+      Object.entries(devDeps).filter(([name]) => {
+        return !name.startsWith("@types/") && name !== "typescript";
+      })
+    ),
+  };
+  // 4. Remove typecheck npm script from package.json
+  if (pkg.scripts && pkg.scripts.typecheck) {
+    delete pkg.scripts.typecheck;
+  }
+  fse.writeJSONSync(path.join(projectDir, "package.json"), newPackageJson, {
+    spaces: 2,
+  });
```

Comment:
Does this reorder `package.json`? E.g. if I had stuff after `devDependencies`, will `...pkg, devDependencies: ...` cause `devDependencies` to show up last? That could create a lot of `git` noise. Maybe mutating `pkg` (as much as I normally prefer immutable stuff) would be better if it preserves key ordering in `package.json`.

Reply:
yeah we probably want to mutate or re-sort it

Reply:
I think we use https://www.npmjs.com/package/sort-package-json in sooner other places